### PR TITLE
Timeline

### PR DESCRIPTION
**Pull request by flexsurfer**
Timeline

### CODE REVIEW

**Comment by qoqobolo** [comment]
### ISSUE 1: White user-pics

_Steps:_
- Open Status tab
- Add new status

User-pic is returned after navigating to another user's profile.

### ISSUE 2: After posting the fist status and re-login in Timeline appear the statuses of a user who has not been added to contacts
After one more re-login they disappear. 

_Steps:_
- Login
- Post the status
- Relogin
- Open the Status tab
- Relogin

Screenshot 2020-11-04 at 11 39 38

Android, iOS

**Comment by churik** [comment]
## ISSUE 3: redirect to main profile view after scanning QR code in Profile > Contacts

**Steps:**
- open Profile > Contacts
- Add new contact > Scan QR

**Expected result:**
on Contacts view

**Actual result:**
redirected to profile view
**Build:** commit https://github.com/status-im/status-react/commit/db78e78c5af3af48187c8c6862431fe5e680ae41

## ISSUE 4: after adding contact no statuses from him are shown in Status until opening his profile or relogin

**Steps:**
- User A: set new status
- User B: add user A as contact (don't open his profile)
- User B: open "Status"

**Expected result:**
- User B: status of User A is shown in Status

**Actual result:**
no status until relogin or opening User A profile
**Build:** commit https://github.com/status-im/status-react/commit/db78e78c5af3af48187c8c6862431fe5e680ae41

## ISSUE 5: no autoscrolling when message is too long (overlapped with keyboard, can't see what I'm typing)
**Steps:**
- User A: open my status, type a message that contains > 20 rows
Monosnap 2020-11-04 12-10-35

## ISSUE 6: last status is erased in timeline if switching to public chat view and back

**Steps:**
- User B: join public channel, open chat view
- User B: add user A as contact, open his profile (to avoid issue 4), open "Status"
- User A: set new status
- User B: switch to home and switch back to Status

**Expected result:**
can see last status

**Actual result:**
last status is not shown; after reopening "Status" again it is shown

**Comment by flexsurfer** [comment]
ISSUE 5: @errorists its kinda tricky to do because of image , what behaviour would you expect ?

**Comment by flexsurfer** [comment]
ISSUE 2: magic , can't reproduce :(
ISSUE 4: can't reproduce :(

**Comment by errorists** [comment]
re: issue 5, I would expect the view to scroll :) [here's a video ](https://www.dropbox.com/s/cdrst0njln3ume0/ScreenFlow.mp4?dl=0)

edit: reuploaded with the autoscroll visible

**Comment by churik** [comment]
@flexsurfer 
Re: issue 2: seems that last update from this user was added 23H ago, can you post new one to see that it is not reproducible for sure? (as far as I understand it is your user)
Re: issue 4: I added contact via Profile > Contacts.
It is not reproducible if you add Contact via + > Scan QR

**Comment by churik** [comment]
## Question 6: why fr own status when you have 3-random name there is no indicator "You", but with ENS it is shown?
@errorists need your opinion. I would prefer to see where are my messages in a more explicit way then hightlighting

photo_2020-11-04 12 56 37 jpeg 2020-11-04 12-57-04

**Comment by errorists** [comment]
that 'you' indicator is not necessary yet :) tl;dr we're discussing adding your Chat Key next to every instance of username in chats, including here. So basically it would be Chat key for everyone else and 'You' in its place for your own updates. until that is live, we can remove that

edit: here's a screenshot to illustrate what I mean, that grey part next to usernames is not necessary yet. Apologies for any confusion!

Screenshot 2020-11-04 at 13 01 39

**Comment by flexsurfer** [comment]
yeah i know issue 4: is not reproducible in any case

**Comment by flexsurfer** [comment]
so should i remove `You` or ?

**Comment by errorists** [comment]
Yes please 🙏  I'm removing it from Figma too

**Comment by flexsurfer** [comment]
thanks @churik @qoqobolo 1,2,3,5 fixed,   4,6 - can't reproduce, looking

**Comment by flexsurfer** [comment]
issue 4 should be fixed

**Comment by flexsurfer** [comment]
issue 6 fixed

**Comment by churik** [comment]
thank you @flexsurfer !
All 6 issues are fixed.
### ISSUE 7: no "fetch more messages" button until reopening  public chat view

**Steps:**
- join empty chat

**Expected result:**
visible 'Fetch more messages'

**Actual result:**
IMAGE 2020-11-05 09:35:12
After reopning chat view "Fetch more messages" is visible

### ISSUE 8: updates from contacts will not appear in timeline on synced device until opening user profile of new contact

**Steps:**
- User A: restore same account on devices 1 and 2, pair devices and sync them
- User A: add contact B with status updates on device 1. Status from User B will be shown on "Status"
- User A: open "Status" on device 2 (contact B is synced, so it is shown in Contacts in Profile)

**Expected result:** can see "Status" of User B in timeline on device 2

**Actual result:** no updates from User B until manual opening his profile

**Comment by flexsurfer** [comment]
thanks @churik nice catch, fixed

**Comment by churik** [comment]
@flexsurfer  
regarding issue 8:
if user A adds contact on `device 1`, on paired `device 2`popup with error `Object is not a function` is shown
IMAGE 2020-11-05 14:29:04
[Status-debug-logs.zip](https://github.com/status-im/status-react/files/5494679/Status-debug-logs.zip)
On initial device after ~1 min another issue:
IMAGE 2020-11-05 14:30:52

**Comment by churik** [comment]
All issues are fixed.
Tests are fixed in https://github.com/status-im/status-react/pull/11368.

@flexsurfer it is impressive piece of work, thank you!

**Comment by errorists** [comment]
@flexsurfer agree, you rock dude!

File: src/status_im/events.cljs [hunk 1]
```diff
@@ -1249,10 +1255,20 @@
   (let [chat-id (chat/profile-chat-topic public-key)]
     (when-not (= (:current-chat-id db) chat-id)
       (fx/merge cofx
-                (chat/start-public-chat chat-id {:dont-navigate? true :profile-public-key public-key})
+                (chat/start-profile-chat public-key)
                 (chat/offload-all-messages)
                 (chat/preload-chat-data chat-id)))))
 
+(fx/defn reset-current-timeline-chat [{:keys [db] :as cofx}]
+  (let [profile-chats (conj (map :public-key (contact.db/get-active-contacts (:contacts/contacts db))) (get-in db [:multiaccount :public-key]))]
```
Comment (rasom) [inline src/status_im/events.cljs]: This is a very long line.

File: src/status_im/ui/screens/contacts_list/views.cljs [hunk 2]
```diff
@@ -21,50 +21,45 @@
...
+           :chevron             true
+           :accessory           :text
+           :accessory-text      blocked-contacts-count
+           :on-press            #(re-frame/dispatch [:navigate-to :blocked-users-list])}]])
+      (if (seq contacts)
+        [list.views/flat-list
+         {:data                      contacts
+          :key-fn                    :address
+          :render-fn                 contacts-list-item}]
+        [react/view {:align-items :center :flex 1 :justify-content :center}
```
Comment (rasom) [inline src/status_im/ui/screens/contacts_list/views.cljs]: commas or pair per line please

File: src/status_im/ui/screens/status/new/views.cljs [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+                                 :margin-right     4
+                                 :border-radius    4})
+                 :source {:uri uri}}]])
+
+(defview photos []
+  (letsubs [camera-roll-photos [:camera-roll-photos]]
+    {:component-did-mount #(re-frame/dispatch [:chat.ui/camera-roll-get-photos 20])}
+    [react/scroll-view {:horizontal                   true :style {:max-height 88}
+                        :keyboard-should-persist-taps :handled}
+     [react/view {:height         88 :border-top-width 1 :border-top-color colors/gray-lighter
```
Comment (rasom) [inline src/status_im/ui/screens/status/new/views.cljs]: extremely hard to read, please align all maps in PR

File: src/status_im/ui/screens/status/new/views.cljs [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                :after               :main-icon/send
+                :disabled            (and (string/blank? @input-text) (not uri))
+                :on-press            #(do
+                                        (re-frame/dispatch [:profile.ui/send-my-status-message])
+                                        (re-frame/dispatch [:navigate-back]))}
+               (i18n/label :t/wallet-send)]}]
+            [react/view {:top      0 :bottom 0 :left 0 :right 0 :align-items :center :justify-content :center
+                         :position :absolute :pointerEvents :none}
+             [react/text {:style {:color colors/gray}}
+              (str (count @input-text) " / 300")]]]]]]))))
```
Comment (rasom) [inline src/status_im/ui/screens/status/new/views.cljs]: what's `" / 300"`? Should it go to translations or something?

File: src/status_im/ui/screens/status/views.cljs [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            [status-im.ui.components.plus-button :as components.plus-button]
+            [status-im.ui.screens.chat.photos :as photos]
+            [status-im.ui.components.tabs :as tabs]
+            [status-im.utils.contenthash :as contenthash]
+            [status-im.multiaccounts.core :as multiaccounts]))
+
+(defonce messages-list-ref (atom nil))
+
+(defn message-content-image [uri _]
+  (let [dimensions (reagent/atom [260 260])]
```
Comment (rasom) [inline src/status_im/ui/screens/status/views.cljs]: can we extract/name magic 260 value?

File: src/status_im/ui/screens/status/views.cljs [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+      [react/text {:style {:font-size 10 :color colors/gray}}
+       (datetime/time-ago (datetime/to-date timestamp))]]
+     (if (= content-type constants/content-type-image)
+       [image-message message]
+       [message/render-parsed-text (assoc message :outgoing false) (:parsed-text content)])]]])
+
+(defn render-message [timeline? account]
+  (fn [{:keys [type] :as message} idx]
+    (if (= type :datemark)
+      [react/view]
```
Comment (rasom) [inline src/status_im/ui/screens/status/views.cljs]: just `nil` will not work here?

File: src/status_im/ui/screens/status/views.cljs [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        account @(re-frame/subscribe [:multiaccount])]
+    [react/view {:flex 1}
+     ;;TODO implement in the next iteration
+     #_[tabs]
+     [react/view {:height 1 :background-color colors/gray-lighter}]
+     (if no-messages?
+       [react/view {:padding-horizontal 32 :margin-top 64}
+        [react/image {:style  {:width 140 :height 140 :align-self :center}
+                      :source {:uri (contenthash/url
+                                     "e3010170122080c27fe972a95dbb4b0ead029d2c73d18610e849fac197e91068a918755e21b2")}}]
```
Comment (rasom) [inline src/status_im/ui/screens/status/views.cljs]: constant?

File: src/status_im/ui/screens/status/views.cljs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+          (if (= :timeline (:tab @state))
+            (i18n/label :t/statuses-descr)
+            (i18n/label :t/statuses-my-status-descr))]]]
+       [list/flat-list
+        {:key-fn                    #(or (:message-id %) (:value %))
+         :render-fn                 (render-message (= :timeline (:tab @state)) account)
+         :data                      messages
+         :on-viewable-items-changed chat.views/on-viewable-items-changed
+         :on-end-reached            #(re-frame/dispatch [:chat.ui/load-more-messages])
+         :on-scroll-to-index-failed #()                     ;;don't remove this
```
Comment (rasom) [inline src/status_im/ui/screens/status/views.cljs]: don't remove which line?

**Comment by flexsurfer** [comment]
thanks @rasom , and sorry about maps, all should be fixed now

**Comment by churik** [comment]
@flexsurfer 
### ISSUE 9: sent messages reappear after relogin if history  was cleared
if you send several messages to public chat, then clear history - messages will reappear after re-login. 
This is the reason of failed e2e.

**Comment by churik** [comment]
### ISSUE 10: After setting status can't  login: got "Invalid chat" type

Can't login anymore.
Steps:
- create account
- set status
- reopen app and try to login
IMAGE 2020-11-06 10:32:39
LOGS: [status_logs.zip](https://github.com/status-im/status-react/files/5500356/status_logs.zip)

**Comment by rasom** [comment]
ISSUE 9
this one is mine, working on it

**Comment by flexsurfer** [comment]
@churik thanks 10 fixed

**Review by rasom — APPROVED** [review]

**Comment by churik** [comment]
Ready to merge @flexsurfer

**Pull request merged**